### ISSUE TITLE
parallel solve for GA

### ISSUE THREAD

**Issue by 2sn** [conradtchan/starfit#107]
Potentially useful for large populations (recommended).

**Comment by conradtchan** [comment]
It's unclear if rewriting the genetic algorithm in Fortran will provide significant performance improvements.

I did some quick profiling of `starfit.Ga` on my laptop:
```
starfit.Ga(
    "HE1327-2326.dat",
    "znuc.S4.star.el.y.stardb.gz",
    silent=True,
    time_limit=10,
    sol_size=3,
    z_max=30,
)
```
92% of the time spent in a GA step is already spent calling the Fortran fitness function. When increasing the population size from the default value of 200 to 2000, this goes up to 96%. This might be because the GA operations that are currently done with NumPy methods scale more efficiently. These parts of the code account for a smaller amount of run time as population size increases.

**Comment by 2sn** [comment]
The questions is how much of this overhead is spent on the interface as compared to actual calculations.  For my MultiStar code the f2py overhead was a factor 50 (when I used my FORTRAN integrator calling FORTRAN functions over SciPy integrator calling FORTRAN functions).  This is what I wanted to save as well.

It is good to see - although admittedly a bit surprising - that python GA code portion runs so swiftly.  
How was this assessed?  No other Numpy calls counting toward it?

**Comment by 2sn** [comment]
BTW, you did run the timing for all the cases, did you check the results?  
I am still quite uncertain how the thread-private data is set initially for each thread.

**Comment by conradtchan** [comment]
> How was this assessed? No other Numpy calls counting toward it?

I profiled it using pyinstrument, which generates a nice timing tree:
```
pyinstrument -r html test.py
```

test.py:
```
import starfit

test_result = starfit.Ga(
    "HE1327-2326.dat",
    "znuc.S4.star.el.y.stardb.gz",
    silent=True,
    time_limit=10,
    sol_size=3,
    z_max=30,
)
```

**Comment by conradtchan** [comment]
> BTW, you did run the timing for all the cases, did you check the results?

The test suite passes for all cases.



> I am still quite uncertain how the thread-private data is set initially for each thread.

The threadprivate variables do not need to be initialised outside of the parallel region. They are always written to prior to calling the solver inside the parallel region. The different versions of these variables never need to be exchanged across threads.

**Comment by 2sn** [comment]
OK, so it I understand correctly, the thread-private variables inherit their values -- are initialised -- from the serial section then.  Is this a documented feature?  I will have to re-read and understand the structure of the parallel code ... thank for all your work.

**Comment by conradtchan** [comment]
No, I don't think that's correct - they don't inherit their values from the serial section. In StarFit, the values are always set inside the parallel region. Prior to that, they are undefined.

See the [OpenMP docs](https://www.openmp.org/spec-html/5.0/openmpsu104.html). In particular:
> Each copy of a threadprivate variable is initialized once, in the manner specified by the program, but at an unspecified point in the program prior to the first reference to that copy.

**Comment by 2sn** [comment]
So, what does this mean "specified by the program"?
I need to have a closer look at the code to see what is done ...
